packages/braintree-integration/src/braintree-fastlane/braintree-fastlane-utils.ts

```diff
@@ -0,0 +1,541 @@
+                });
+
+                this.browserStorage.setItem('sessionId', cart.id);
+
+                return;
+            }
+
+            const { authenticationState, profileData } = await triggerAuthenticationFlow(
+                customerContextId,
+            );
+
+            if (authenticationState === BraintreeFastlaneAuthenticationState.CANCELED) {
+                await this.paymentIntegrationService.updatePaymentProviderCustomer({
+                    authenticationState,
+                    addresses: [],
+                    instruments: [],
+                });
+
+                this.browserStorage.removeItem('sessionId');
+
+                return;
+            }
+
+            const shippingAddresses =
+                this.mapPayPalFastlaneToBcAddress(profileData.shippingAddress) || [];
+            const paypalBillingAddresses = this.getPayPalFastlaneBillingAddresses(profileData);
+            const billingAddresses =
+                this.mapPayPalFastlaneToBcAddress(paypalBillingAddresses) || [];
+            const instruments =
+                this.mapPayPalFastlaneToBcInstrument(methodId, profileData.card) || [];
+            const addresses = this.mergeShippingAndBillingAddresses(
+                shippingAddresses,
+                billingAddresses,
+            );
+
+            this.browserStorage.setItem('sessionId', cart.id);
+            await this.paymentIntegrationService.updatePaymentProviderCustomer({
+                authenticationState,
+                addresses,
+                instruments,
+            });
+
+            if (billingAddresses.length > 0) {
+                await this.paymentIntegrationService.updateBillingAddress(billingAddresses[0]);
+            }
+
+            if (shippingAddresses.length > 0 && cart.lineItems.physicalItems.length > 0) {
+                await this.paymentIntegrationService.updateShippingAddress(shippingAddresses[0]);
+            }
+        } catch (error) {
+            // TODO: we should figure out what to do here
+            // TODO: because we should not to stop the flow if the error occurs on paypal side
+        }
+    }
+
+    /**
+     *
+     * PayPal to BC data mappers
+     *
+     * */
+    private mapPayPalToBcAddress(
+        addresses?: BraintreeConnectAddress[],
+        phones?: BraintreeConnectPhone[],
+    ): CustomerAddress[] | undefined {
+        if (!addresses) {
+            return;
+        }
+
+        const countries = this.paymentIntegrationService.getState().getCountries() || [];
+        const phoneNumber =
+            phones && phones[0] ? phones[0].country_code + phones[0].national_number : '';
+
+        const getCountryNameByCountryCode = (countryCode: string) => {
+            const matchedCountry = countries.find((country) => country.code === countryCode);
+
+            return matchedCountry?.name || '';
+        };
+
+        return addresses.map((address) => ({
+            id: Number(address.id),
```

Comment:
Can you double check Braintree Fastlane 0.7 integration guide to be sure that PP returns address.id in payload?

Reply:
i created different method for BT fastlane `mapPayPalFastlaneToBcAddress` that's why i created `runPayPalFastlaneAuthenticationFlowOrThrow` , because there are several methods that are differnt from connect

Reply:
Please, take a look on PayPal Commerce Fastlane address mapper example. As far as I know, Braintree Fastlane will not return address id, so we should update it with something unique instead of this